### PR TITLE
Add support for type-variant generics

### LINKED ISSUE #140: Support for Mapping, MutableMapping, Sequence, etc?
Marshmallow supports serializing arbitrary sequences and mappings, so long as they provide the correct interface. Unfortunately, desert doesn't seem to allow generic types out of the box:

The following crashes:

```py
import dataclasses

@dataclasses.dataclass
class D:
    foo: Sequence[str]

schema = desert.schema(D)
```

Traceback:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dargueta/.pyenv/versions/3.8.6/lib/python3.8/site-packages/desert/__init__.py", line 24, in schema
    return desert._make.class_schema(cls, meta=meta)(many=many)
  File "/Users/dargueta/.pyenv/versions/3.8.6/lib/python3.8/site-packages/desert/_make.py", line 130, in class_schema
    attributes[field.name] = field_for_schema(
  File "/Users/dargueta/.pyenv/versions/3.8.6/lib/python3.8/site-packages/desert/_make.py", line 291, in field_for_schema
    nested = forward_reference or class_schema(typ)
  File "/Users/dargueta/.pyenv/versions/3.8.6/lib/python3.8/site-packages/desert/_make.py", line 105, in class_schema
    raise desert.exceptions.UnknownType(
desert.exceptions.UnknownType: Desert failed to infer the field type for typing.Sequence[str].
Explicitly pass a Marshmallow field type.
```

Can we add support for these? It seems like the changes would need to be made [here](https://github.com/python-desert/desert/blob/main/src/desert/_make.py#L241-L257) and not too much work.

### Environment

Python 3.8.6

* desert: 2020.11.18
* marshmallow: 3.13.0
* marshmallow-enum: 1.5.1
* marshmallow-union: 0.1.15.post1

### PR DESCRIPTION
**Pull request by dargueta**
Closes #140

I see the contributing guidelines say to add a file to `changelog.d` but the directory appears to be empty aside from the towncrier template... are the contributing docs out of date?

### CODE REVIEW

**Comment by dargueta** [comment]
Ah okay I'll create the file then. Thanks for clarifying.

File: src/desert/_make.py
Comment (altendky) [inline src/desert/_make.py]: I wouldn't come up with an abbreviation.

File: tests/test_make.py
Comment (altendky) [inline tests/test_make.py]: (applies to multiple locations)

https://docs.python.org/3.10/library/functions.html#classmethod
> `def f(cls, arg1, arg2, ...): ...`

https://www.python.org/dev/peps/pep-0008/#designing-for-inheritance
> 'cls' is the preferred spelling for any variable or argument which is known to be a class

```suggestion
    cls = type("A", (object,), {"__annotations__": {"y": annotation_class[int]}})
```

File: tests/test_make.py
Comment (altendky) [inline tests/test_make.py]: This seems like it is about dropping 3.6 support?  Maybe?  That would be a separate PR.  Or, maybe I'm misreading this.

**Review by altendky — CHANGES_REQUESTED** [review]
Given the CI failures, do we need to drop 3.6 support first?  Or xfail these new cases on 3.6?  Without saying it should be the latter, here's an example of how.

```python
@pytest.mark.parametrize(
    argnames="an_arg",
    argavalues=[
        pytest.param(
            typing.Sequence,
            marks=pytest.mark.xfail(condition=sys.version_info < (3, 7), reason="3.6 yada yada"),
        ),
    ],
)
```

File: tests/test_make.py
Comment (dargueta) [reply tests/test_make.py]: Nope, that was an accident. I'll fix it.

Comment (dargueta) [reply tests/test_make.py]: We're adding support for `Sequence` so I figured we needed to change it to something else. I didn't realize that would break 3.6.

Comment (altendky) [reply tests/test_make.py]: Alrighty, changing the bit inside the test makes sense but i think the skip ends up still applying?

File: tests/test_make.py
Comment (altendky) [inline tests/test_make.py]: Let's avoid the need for this in the future by just creating our own `Generic`.

```python3
T = t.TypeVar("T")
class C(t.Generic[T]):
    pass

@module.dataclass
class A:
    x: C[int]
```

File: changelog.d/140.change.rst
Comment (altendky) [inline changelog.d/140.change.rst]: Doesn't it change behavior though?  Previously you couldn't build a schema with hints of these types and now you can.

**Review by altendky — CHANGES_REQUESTED** [review]
Getting close... thanks again.

File: changelog.d/140.change.rst
Comment (dargueta) [reply changelog.d/140.change.rst]: Poor phrasing, agreed. How about changing it to "validation behavior"?

File: tests/test_make.py
Comment (dargueta) [reply tests/test_make.py]: Maybe I'm misunderstanding this -- what exactly is this skip avoiding?

Comment (altendky) [reply tests/test_make.py]: I don't really remember so let me write out my exploration here.  To start with, in 3.6 we get:

https://github.com/python-desert/desert/runs/4645282988?check_suite_focus=true#step:5:147
```python-traceback
______________________ test_raise_unknown_generic[attrs] _______________________
tests/test_make.py:543: in test_raise_unknown_generic
    desert.schema_class(A)
.tox/py36/lib/python3.6/site-packages/desert/__init__.py:47: in schema_class
    return desert._make.class_schema(cls, meta=meta)
.tox/py36/lib/python3.6/site-packages/desert/_make.py:130: in class_schema
    field.metadata,
.tox/py36/lib/python3.6/site-packages/desert/_make.py:313: in field_for_schema
    nested = forward_reference or class_schema(typ)
.tox/py36/lib/python3.6/site-packages/desert/_make.py:120: in class_schema
    raise desert.exceptions.NotAnAttrsClassOrDataclass(clazz)
E   desert.exceptions.NotAnAttrsClassOrDataclass: tests.test_make.UnknownGeneric[int]
```
```python-traceback
___________________ test_raise_unknown_generic[dataclasses] ____________________
tests/test_make.py:543: in test_raise_unknown_generic
    desert.schema_class(A)
.tox/py36/lib/python3.6/site-packages/desert/__init__.py:47: in schema_class
    return desert._make.class_schema(cls, meta=meta)
.tox/py36/lib/python3.6/site-packages/desert/_make.py:130: in class_schema
    field.metadata,
.tox/py36/lib/python3.6/site-packages/desert/_make.py:313: in field_for_schema
    nested = forward_reference or class_schema(typ)
.tox/py36/lib/python3.6/site-packages/desert/_make.py:120: in class_schema
    raise desert.exceptions.NotAnAttrsClassOrDataclass(clazz)
E   desert.exceptions.NotAnAttrsClassOrDataclass: tests.test_make.UnknownGeneric[int]
```

It was originally added in https://github.com/python-desert/desert/commit/d61f8db212112d5c40a587252dd558826e558fb1 and relates to the following code.

https://github.com/python-desert/desert/blob/32d4f6863886b8628ad0dcd670d51f69b6dd5c3b/src/desert/_make.py#L104-L117

Python 3.6 does treat the `isinstance()` check differently.

```console
$ for v in 3.6 3.7 3.8 3.9 3.10; do python$v -c "import typing; print('$v', isinstance(typing.Sequence, type))"; done
3.6 True
3.7 False
3.8 False
3.9 False
3.10 False
```

I haven't thought through the implications or how we might work around this to fix 3.6, but given that 3.6 is EOL now I think just adding the skip back until we remove 3.6 is probably the easiest path forward.

Once you get one submission merged GitHub will run builds for you without approval so you can see this stuff without waiting for our approval.  I guess awhile back they started running into too many PRs that were doing cryptocurrency mining or something so they locked down first-time contributors from having their builds run automatically.

Comment (dargueta) [reply tests/test_make.py]: > I haven't thought through the implications or how we might work around this to fix 3.6, but given that 3.6 is EOL now I think just adding the skip back until we remove 3.6 is probably the easiest path forward.

Sounds good.

File: tests/test_make.py
Comment (altendky) [inline tests/test_make.py]:
```suggestion
    cls = type("A", (object,), {"__annotations__": {"y": annotation_class[int]}})  # type: ignore[index]
```

File: tests/test_make.py
Comment (altendky) [inline tests/test_make.py]:
```suggestion
    cls = type("A", (object,), {"__annotations__": {"y": annotation_class[int, int]}})  # type: ignore[index]
```

**Review by altendky — CHANGES_REQUESTED** [review]
https://github.com/python-desert/desert/runs/4655242060?check_suite_focus=true#step:5:266

```
tests/test_make.py:169: error: Value of type "type" is not indexable
tests/test_make.py:180: error: Value of type "type" is not indexable
```

So I was able to satisfy mypy, but I think that just ignoring is better in this case.  The solution was super specific and overly verbose etc.  I think there isn't a proper way to write a hint for `a generic with a single typevar'.

**Review by altendky — APPROVED** [review]
Thanks for seeing this through.  Sorry for the wait.

**Pull request merged**
